parquet-variant-compute/src/unshred_variant.rs

```diff
@@ -0,0 +1,520 @@
+    }
+
+    // NOTE: None/None at top-level is technically invalid, but the shredding spec requires us to
+    // emit `Variant::Null` when a required value is missing.
+    let mut row_builder = make_unshred_variant_row_builder(array.shredding_state().borrow())?
+        .unwrap_or_else(|| UnshredVariantRowBuilder::null(array.nulls()));
+
+    let metadata = array.metadata_field();
+    let mut value_builder = VariantValueArrayBuilder::new(array.len());
+    let mut null_builder = NullBufferBuilder::new(array.len());
+    for i in 0..array.len() {
+        if array.is_null(i) {
+            value_builder.append_null();
+            null_builder.append_null();
+        } else {
+            let metadata = VariantMetadata::new(metadata.value(i));
+            let mut value_builder = value_builder.builder_ext(&metadata);
+            row_builder.append_row(&mut value_builder, &metadata, i)?;
+            null_builder.append_non_null();
+        }
+    }
+
+    let value = value_builder.build()?;
+    Ok(VariantArray::from_parts(
+        metadata.clone(),
+        Some(value),
+        None,
+        null_builder.finish(),
+    ))
+}
+
+/// Row builder for converting shredded VariantArray rows back to unshredded form
+enum UnshredVariantRowBuilder<'a> {
+    PrimitiveInt8(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Int8Type>>),
+    PrimitiveInt16(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Int16Type>>),
+    PrimitiveInt32(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Int32Type>>),
+    PrimitiveInt64(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Int64Type>>),
+    PrimitiveFloat32(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Float32Type>>),
+    PrimitiveFloat64(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Float64Type>>),
+    PrimitiveDate32(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Date32Type>>),
+    PrimitiveTime64(UnshredPrimitiveRowBuilder<'a, PrimitiveArray<Time64MicrosecondType>>),
+    TimestampMicrosecond(TimestampUnshredRowBuilder<'a, TimestampMicrosecondType>),
+    TimestampNanosecond(TimestampUnshredRowBuilder<'a, TimestampNanosecondType>),
+    PrimitiveBoolean(UnshredPrimitiveRowBuilder<'a, BooleanArray>),
+    PrimitiveString(UnshredPrimitiveRowBuilder<'a, StringArray>),
+    PrimitiveBinaryView(UnshredPrimitiveRowBuilder<'a, BinaryViewArray>),
+    PrimitiveUuid(UnshredPrimitiveRowBuilder<'a, FixedSizeBinaryArray>),
+    Struct(StructUnshredVariantBuilder<'a>),
+    ValueOnly(ValueOnlyUnshredVariantBuilder<'a>),
+    Null(NullUnshredVariantBuilder<'a>),
+}
+
+impl<'a> UnshredVariantRowBuilder<'a> {
+    /// Creates an all-null row builder.
+    fn null(nulls: Option<&'a NullBuffer>) -> Self {
+        Self::Null(NullUnshredVariantBuilder::new(nulls))
+    }
+
+    /// Appends a single row at the given value index to the supplied builder.
+    fn append_row(
+        &mut self,
+        builder: &mut impl VariantBuilderExt,
+        metadata: &VariantMetadata,
+        index: usize,
+    ) -> Result<()> {
```

Comment:
I don't know how much it matters, but you could avoid the dynamic dispatch here with a generic function:

```suggestion
    fn append_row<T: VariantBuilderExt>(
        &mut self,
        builder: &mut T,
        metadata: &VariantMetadata,
        index: usize,
    ) -> Result<()> {
```

Reply:
🤦  you are right